### ISSUE TITLE
The CSS class `is-dark-theme` is not set in the post editor 

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#60299]
### Description

Splitting this out from https://github.com/WordPress/gutenberg/issues/60275

To be more accurate, in the Post editor the CSS class `is-dark-theme` is initially applied but than removed.
In the Site editor, it works as expected.

It seems to be something missed with the refactorings of the editor irame, editor styles scoping and such.

When a theme has a dark background color, [useDarkThemeBodyClassName](https://github.com/WordPress/gutenberg/blob/e03b68cc5c35c0666281ad7f7dc6474495833e6e/packages/block-editor/src/components/editor-styles/index.js#L24) is responsible for adding the `is-dark-theme` CSS class to the editor iframe body element.

It does that, but then the body gets updated and gets new CSS classes when the iframe is fully loaded, see the `Iframe` component internal handling of the classes in the [onload callback](https://github.com/WordPress/gutenberg/blob/32a90263edaa9f6e6f0a9675722febeb2055143a/packages/block-editor/src/components/iframe/index.js#L153-L160) and then [the classes are set to the body](https://github.com/WordPress/gutenberg/blob/32a90263edaa9f6e6f0a9675722febeb2055143a/packages/block-editor/src/components/iframe/index.js#L362-L373).

I'm not sure whether that means the `body` gets entirely replaced. Regardless the new CSS classes _replace_ the initial ones thus removing `is-dark-theme` it it was set.

The `is-dark-theme` CSS class is necessaty to make some editor features and styles work correctly.

### Step-by-step reproduction instructions

- Set Twenty Twenty-Four as active theme.
- Go to the Site editor > Styles, set a theme style variation that has a dark background e.g. 'Onyx' and save. **Update**: the Site editor UI has changed and now 'Onyx' is shown as a _palette_ variation in the Styles panel > Browse styles
- In your browser dev tools, inspect the editor iframe body ie. the body with CSS classes `block-editor-iframe__body editor-styles-wrapper`.
- Observe the body does have a CSS class `is-dark-theme`.
- Go to the Post editor.
- Observe the editor shows the theme variant with dark background.
- In your browser dev tools, inspect the editor iframe body ie. the body with CSS classes `block-editor-iframe__body editor-styles-wrapper`.
- If you inspect it very quickly, you will notice that initially the body element has these CSS classes: `block-editor-iframe__body editor-styles-wrapper is-dark-theme`.
- When the iframe is fully loaded, observe the element body element does _not_ have the `is-dark-theme` class.
- Observe the body element CSS classes have been replaced with `block-editor-iframe__body editor-styles-wrapper post-type-post admin-color-fresh wp-embed-responsive`.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by afercia** [comment]
The commit to fix this issue was reverted to fix a problem with Safari, see https://github.com/WordPress/gutenberg/pull/60616.

I'm reopening this issue as it needs to be solved, but with a different approach.

**Comment by ellatrix** [comment]
Not sure if that would break anything, but maybe the best wait to fix the issue is to move `is-dark-theme` to the `html` element, which is not managed in React?

**Comment by afercia** [comment]
Based on [this comment](https://github.com/WordPress/gutenberg/pull/60550#discussion_r1556669468), looks like the problem with Safari is broader because also the CSS classes `admin-color-{$color_scheme}` and `wp-embed-responsive` seems to not be set correctly in Safari. However, I can't reproduce that issue on trunk.

Regarding the bug reported at https://github.com/WordPress/gutenberg/pull/60550 I'm not sure the root cause was the change in https://github.com/WordPress/gutenberg/pull/60300 as that change followed an already existing pattern used for the `admin-color-{$color_scheme}` and `wp-embed-responsive` classes.

Anyways, to illustrate a couple of the issues triggered by the missing `is-dark-theme` CSS classe, please see the screenshot below. It's just two examples. there are probably more issues.

A block in 'select mode' now shows a blue outline (left) but it is supposed to show a white outline (dark):

![Screenshot 2024-04-11 at 15 57 25](https://github.com/WordPress/gutenberg/assets/1682452/45cc121e-5599-468c-9258-bdc83844906b)

In addition to the outline, the Spacer block, when selected, shows a full dark background (left) but it is supposed to show a lighter background (right):

![Screenshot 2024-04-11 at 16 58 41](https://github.com/WordPress/gutenberg/assets/1682452/9cd7bcd7-c842-4025-9d4d-ec0c937d4de4)

**Comment by ellatrix** [comment]
My suggestion would be to move the class to the `html` element so that it's not overridden later on

**Comment by afercia** [comment]
> My suggestion would be to move the class to the `html` element so that it's not overridden later on

I've been testing pretty exensively. While moving the class to the thml element seems trivial, it doesn't work well anyways with themes that _may_ handle the body background dynamically. For example, Twenty Twenty-One does have its won 'dark mode toggle' feature that _changes_ the body background _after_ the editor `useDarkThemeBodyClassName` checks the background color luminance. 

That's a pre-existing issue and I suspect it never worked well. However, it seems there's a need to a better mechanism to get the actual body background color even when it gets changed by themes or plugins features.

**Comment by afercia** [comment]
A [first attempt](https://github.com/WordPress/gutenberg/pull/60300) to fix this issue was reverted in https://github.com/WordPress/gutenberg/pull/60616. I can still reproduce the incorrect behavior in the Post editor. @glendaviesnz @ellatrix any change to explore an alternative approach to fix this issue>
RIght now, the implementation of the `is-dark-theme` class in the Post editor is just broken.

**Comment by afercia** [comment]
This issue is still a blocker for https://github.com/WordPress/gutenberg/issues/60275
I would appreciate some focus on it to find a better, more solid, solution than the one tried in the first attempt.
Cc @WordPress/gutenberg-core 

To recap: In the _post_ editor the CSS classes of the iframe body are initially:

`block-editor-iframe__body editor-styles-wrapper is-dark-theme`

and after the page is fully loaded, a re-render replaces the classes with:

`block-editor-iframe__body editor-styles-wrapper post-type-post admin-color-fresh wp-embed-responsive`

As such, `is-dark-theme` is lost and there's no way to distinguish a light theme from a dark theme. This doesn't allow to address the accessibility issue in https://github.com/WordPress/gutenberg/issues/60275

**Comment by afercia** [comment]
Previously related: https://github.com/WordPress/gutenberg/issues/64548 / https://github.com/WordPress/gutenberg/pull/64549

**Comment by richtabor** [comment]
I'm against changing the admin selected UI color as a result of the theme's background color (the block outlines).

It doesn't scale as-is, and actually makes the experience more confusing as what was once familiar is no longer. 

If we were to work on this, the resulting solution needs to be complete, accounting for:
- theme background contrast comparable to the admin selected color. 
- a parent block's background color comparable to the selected color of blocks nested within. 

Otherwise, you can have a black site with white sections and you no longer see any selected blocks. 

It cannot be a site-wide condition.

**Comment by afercia** [comment]
Thanks for your feedback. I'm not sure it's directly related to this issue. It may be a follow-up consideration to address separately.
The fact the `is-dark-theme` CSS class works in the Site editor and doesn't in the Post editor is a bug that must be fixed.
This bug is a blocker for https://github.com/WordPress/gutenberg/issues/60275 which is an accessibility issue that needs to be fixed.

**Comment by colinduwe** [comment]
I propose adding a bodyClasses property to settings in the 'core/block-editor' store and refactoring the iframe component to use that state to maintain its body classes. [useDarkThemeBodyClassName](https://github.com/WordPress/gutenberg/blob/e03b68cc5c35c0666281ad7f7dc6474495833e6e/packages/block-editor/src/components/editor-styles/index.js#L24) could update the store and not be overwritten by the iframe component's onLoad behavior.
Here's a rough proposal related to some other open issues caused by the current way bodyClasses are applied to the iframe component: https://github.com/WordPress/gutenberg/pull/68538

**Comment by colinduwe** [comment]
I updated my PR to include usDarkThemeBodyClassName. I'd appreciate any feedback.

**Comment by afercia** [comment]
> I updated my PR to include usDarkThemeBodyClassName. I'd appreciate any feedback.

Link to the mentioned PR: https://github.com/WordPress/gutenberg/pull/68538

**Comment by afercia** [comment]
Adding tests on https://github.com/WordPress/gutenberg/pull/68481 surfaced that the `is-dark-theme` class isn't added to the Site editor as well, when the editor is in 'view mode' after initial load. This can be tested by setting a dark theme as active theme and by inspecting the editor canvas iframe body i.e. the element with class `block-editor-iframe__body`.

Screenshot:

![Image](https://github.com/user-attachments/assets/60795990-8776-4e13-bc71-e2e0a10f4ed4)

Only when switching the Site editor to edit mode the `is-dark-theme` CSS class is added. https://github.com/WordPress/gutenberg/pull/68481 should fix this as well.
